### PR TITLE
Reset baserel cache on invalid hypertable cache

### LINKED ISSUE #4795: [Bug]: Assertion failure in  ts_hypertable_has_compression_table
This is hit in sanitizer test occasionally. 

Failed run with coredump:
* https://github.com/timescale/timescaledb/actions/runs/3185109585

Query that triggers (not guaranteed):
```
SELECT
    count(*)
FROM
    metrics_compressed_summary
WHERE
    bucket > _timescaledb_internal.to_timestamp(
        _timescaledb_internal.cagg_watermark(
            (SELECT
                mat_hypertable_id
             FROM
                _timescaledb_catalog.continuous_agg
             WHERE
                user_view_name = 'metrics_compressed_summary')
        )
    );
```

Stacktrace:
```
#2  0x0000555ecac25f18 in ExceptionalCondition (conditionName=conditionName@entry=0x7f0774d9c4e0 "ht->fd.compression_state == HypertableCompressionEnabled", errorType=errorType@entry=0x7f0774d99380 "FailedAssertion",
    fileName=fileName@entry=0x7f0774d99480 "/home/runner/work/timescaledb/timescaledb/src/hypertable.c", lineNumber=lineNumber@entry=2909) at assert.c:69
#3  0x00007f0774c1bfe1 in ts_hypertable_has_compression_table (ht=0x6250000c3870) at /home/runner/work/timescaledb/timescaledb/src/hypertable.c:2909
#4  0x00007f0774d06a35 in timescaledb_get_relation_info_hook (root=0x6310000e10c8, relation_objectid=<optimized out>, inhparent=<optimized out>, rel=0x7f077324aac8) at /home/runner/work/timescaledb/timescaledb/src/planner/planner.c:1277
#5  0x0000555eca3090d1 in get_relation_info (root=root@entry=0x6310000e10c8, relationObjectId=<optimized out>, inhparent=inhparent@entry=false, rel=rel@entry=0x7f077324aac8) at plancat.c:482
#6  0x0000555eca324a3f in build_simple_rel (root=root@entry=0x6310000e10c8, relid=relid@entry=4, parent=parent@entry=0x0) at relnode.c:307
#7  0x00007f0774377b48 in decompress_chunk_add_plannerinfo (root=root@entry=0x6310000e10c8, info=info@entry=0x6310000eaa48, chunk=chunk@entry=0x7f077324bce8, chunk_rel=chunk_rel@entry=0x6310000e2fd0,
    needs_sequence_num=needs_sequence_num@entry=false) at /home/runner/work/timescaledb/timescaledb/tsl/src/nodes/decompress_chunk/decompress_chunk.c:1101
#8  0x00007f0774378ad0 in ts_decompress_chunk_generate_paths (root=root@entry=0x6310000e10c8, chunk_rel=chunk_rel@entry=0x6310000e2fd0, ht=ht@entry=0x625000131338, chunk=chunk@entry=0x7f077324bce8)
    at /home/runner/work/timescaledb/timescaledb/tsl/src/nodes/decompress_chunk/decompress_chunk.c:391
#9  0x00007f0774249980 in tsl_set_rel_pathlist_query (root=0x6310000e10c8, rel=0x6310000e2fd0, rti=<optimized out>, rte=<optimized out>, ht=0x625000131338) at /home/runner/work/timescaledb/timescaledb/tsl/src/planner.c:114
#10 0x00007f0774d0a8a6 in apply_optimizations (root=root@entry=0x6310000e10c8, reltype=reltype@entry=TS_REL_CHUNK_CHILD, rel=rel@entry=0x6310000e2fd0, rte=rte@entry=0x6310000e6698, ht=0x625000131338)
    at /home/runner/work/timescaledb/timescaledb/src/planner/planner.c:1072
#11 0x00007f0774d0b864 in timescaledb_set_rel_pathlist (root=0x6310000e10c8, rel=<optimized out>, rti=<optimized out>, rte=0x6310000e6698) at /home/runner/work/timescaledb/timescaledb/src/planner/planner.c:1219
#12 0x00007f0774cbe64e in set_rel_pathlist (root=root@entry=0x6310000e10c8, rel=rel@entry=0x6310000e2fd0, rti=rti@entry=2, rte=rte@entry=0x6310000e6698) at /home/runner/work/timescaledb/timescaledb/src/import/allpaths.c:251
#13 0x00007f0774cbf624 in ts_set_append_rel_pathlist (root=root@entry=0x6310000e10c8, rel=rel@entry=0x6310000e2b58, rti=rti@entry=1, rte=rte@entry=0x6310000e0028) at /home/runner/work/timescaledb/timescaledb/src/import/allpaths.c:176
#14 0x00007f0774d09a7e in reenable_inheritance (root=root@entry=0x6310000e10c8, rel=rel@entry=0x6310000e2b58, rti=rti@entry=1, rte=rte@entry=0x6310000e0028) at /home/runner/work/timescaledb/timescaledb/src/planner/planner.c:1043
#15 0x00007f0774d0b8bf in timescaledb_set_rel_pathlist (root=0x6310000e10c8, rel=<optimized out>, rti=<optimized out>, rte=0x6310000e0028) at /home/runner/work/timescaledb/timescaledb/src/planner/planner.c:1173
#16 0x0000555eca136c82 in set_rel_pathlist (root=root@entry=0x6310000e10c8, rel=rel@entry=0x6310000e2b58, rti=rti@entry=1, rte=0x6310000e0028) at allpaths.c:541
#17 0x0000555eca137176 in set_base_rel_pathlists (root=root@entry=0x6310000e10c8) at allpaths.c:353
#18 0x0000555eca1391c7 in make_one_rel (root=root@entry=0x6310000e10c8, joinlist=joinlist@entry=0x6310000e3f50) at allpaths.c:223
#19 0x0000555eca20d65e in query_planner (root=root@entry=0x6310000e10c8, qp_callback=qp_callback@entry=0x555eca20f4d8 <standard_qp_callback>, qp_extra=qp_extra@entry=0x7ffe76ca1e40) at planmain.c:276
#20 0x0000555eca231ca8 in grouping_planner (root=root@entry=0x6310000e10c8, tuple_fraction=<optimized out>, tuple_fraction@entry=0) at planner.c:1442
#21 0x0000555eca239bef in subquery_planner (glob=<optimized out>, parse=parse@entry=0x6310000dfeb8, parent_root=parent_root@entry=0x62d000141650, hasRecursion=hasRecursion@entry=false, tuple_fraction=tuple_fraction@entry=0)
    at planner.c:1019
#22 0x0000555eca12f2d5 in set_subquery_pathlist (root=root@entry=0x62d000141650, rel=rel@entry=0x6310000dea68, rti=rti@entry=4, rte=rte@entry=0x62900016bc48) at allpaths.c:2229
#23 0x0000555eca13014d in set_rel_size (root=root@entry=0x62d000141650, rel=rel@entry=0x6310000dea68, rti=rti@entry=4, rte=rte@entry=0x62900016bc48) at allpaths.c:423
#24 0x0000555eca13112a in set_append_rel_size (root=root@entry=0x62d000141650, rel=rel@entry=0x6310000de598, rti=rti@entry=1, rte=rte@entry=0x625000249140) at allpaths.c:1105
#25 0x0000555eca12ff43 in set_rel_size (root=root@entry=0x62d000141650, rel=rel@entry=0x6310000de598, rti=rti@entry=1, rte=rte@entry=0x625000249140) at allpaths.c:384
#26 0x0000555eca132bf7 in set_base_rel_sizes (root=root@entry=0x62d000141650) at allpaths.c:324
#27 0x0000555eca138e73 in make_one_rel (root=root@entry=0x62d000141650, joinlist=joinlist@entry=0x6310000de940) at allpaths.c:186
#28 0x0000555eca20d65e in query_planner (root=root@entry=0x62d000141650, qp_callback=qp_callback@entry=0x555eca20f4d8 <standard_qp_callback>, qp_extra=qp_extra@entry=0x7ffe76ca2650) at planmain.c:276
#29 0x0000555eca231ca8 in grouping_planner (root=root@entry=0x62d000141650, tuple_fraction=<optimized out>, tuple_fraction@entry=0) at planner.c:1442
#30 0x0000555eca239bef in subquery_planner (glob=glob@entry=0x62900016bac8, parse=parse@entry=0x625000006f80, parent_root=parent_root@entry=0x0, hasRecursion=hasRecursion@entry=false, tuple_fraction=tuple_fraction@entry=0)
    at planner.c:1019
#31 0x0000555eca23b3e5 in standard_planner (parse=parse@entry=0x625000006f80,
    query_string=query_string@entry=0x625000005220 "SELECT\n    count(*)\nFROM\n    metrics_compressed_summary\nWHERE\n    bucket > _timescaledb_internal.to_timestamp(\n        _timescaledb_internal.cagg_watermark(\n            (SELECT\n
                mat_hypertable_id\n             FROM\n                _timescaledb_catalog.continuous_agg\n             WHERE\n                user_view_name = 'metrics_compressed_summary')\n        )\n    );",
    cursorOptions=cursorOptions@entry=2048, boundParams=boundParams@entry=0x0) at planner.c:400
#32 0x00007f0774d04892 in timescaledb_planner (parse=<optimized out>, query_string=<optimized out>, cursor_opts=<optimized out>, bound_params=<optimized out>) at /home/runner/work/timescaledb/timescaledb/src/planner/planner.c:578
#33 0x0000555eca23e7e5 in planner (parse=parse@entry=0x625000006f80,
    query_string=query_string@entry=0x625000005220 "SELECT\n    count(*)\nFROM\n    metrics_compressed_summary\nWHERE\n    bucket > _timescaledb_internal.to_timestamp(\n        _timescaledb_internal.cagg_watermark(\n            (SELECT\n
                mat_hypertable_id\n             FROM\n                _timescaledb_catalog.continuous_agg\n             WHERE\n                user_view_name = 'metrics_compressed_summary')\n        )\n    );",
    cursorOptions=cursorOptions@entry=2048, boundParams=boundParams@entry=0x0) at planner.c:269
#34 0x0000555eca6a2856 in pg_plan_query (querytree=querytree@entry=0x625000006f80,
    query_string=query_string@entry=0x625000005220 "SELECT\n    count(*)\nFROM\n    metrics_compressed_summary\nWHERE\n    bucket > _timescaledb_internal.to_timestamp(\n        _timescaledb_internal.cagg_watermark(\n            (SELECT\n
                mat_hypertable_id\n             FROM\n                _timescaledb_catalog.continuous_agg\n             WHERE\n                user_view_name = 'metrics_compressed_summary')\n        )\n    );",
    cursorOptions=cursorOptions@entry=2048, boundParams=boundParams@entry=0x0) at postgres.c:846
#35 0x0000555eca6a30e3 in pg_plan_queries (querytrees=0x6290001686f0,
```
Extra context:
```
(gdb) p *ht
$3 = {fd = {id = 2139062143, schema_name = {
      data = "\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\17
7\177\177\177\177\177\177\177\177"}, table_name = {
      data = "\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\17
7\177\177\177\177\177\177\177\177"}, associated_schema_name = {
      data = "\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\17
7\177\177\177\177\177\177\177\177"}, associated_table_prefix = {
      data = "\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\17
7\177\177\177\177\177\177\177\177"}, num_dimensions = 32639, chunk_sizing_func_schema = {
      data = "\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\17
7\177\177\177\177\177\177\177\177"}, chunk_sizing_func_name = {
      data = "\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\177\17
7\177\177\177\177\177\177\177\177"}, chunk_target_size = 9187201950435737471, compression_state = 32639, compressed_hypertable_id = 2139062143, replication_factor = 32639}, main_table_relid = 2139062143, chunk_sizing_func = 2139062143,
  space = 0x7f7f7f7f7f7f7f7f, chunk_cache = 0x7f7f7f7f7f7f7f7f, data_nodes = 0x7f7f7f7f7f7f7f7f}
```

### PR DESCRIPTION
**Pull request by mkindahl**
When popping the hypertable cache stack, it might happen that the hypertable cache was invalidated between the push and the pop. In that case, the baserel cache can contain invalid entries pointing to the now popped hypertable cache, so we reset the baserel cache.

Fixes #4795

### CODE REVIEW

**Review by akuzm — APPROVED** [review]
I think this works as a quick fix. Would be very good to have a repro of this problem.

File: src/planner/planner.c [hunk 1]
```diff
@@ -225,10 +225,16 @@ planner_hcache_pop(bool release)
...
+	planner_hcaches = list_delete_first(planner_hcaches);
+
 	if (release)
+	{
 		ts_cache_release(hcache);
-
-	planner_hcaches = list_delete_first(planner_hcaches);
+		/* If we pop a stack and discover a new hypertable cache, the basrel
+		 * cache can contain invalid entries, so we reset it. */
+		if (planner_hcaches && hcache != linitial(planner_hcaches))
```
Comment (jnidzwetzki) [inline src/planner/planner.c:235]: Nit: I would compare `planner_hcaches` with `NIL` here.

`linitial` calls `list_nth_cell` which performs an `Assert(list != NIL);`.

```suggestion
		if (planner_hcaches != NIL && hcache != linitial(planner_hcaches))
```

**Review by jnidzwetzki — APPROVED** [review]
The changes look good to me. 

The changes in `.github/workflows/sanitizer-build-and-test.yaml` should be reverted before merging to prevent the OOMs in our regular CI runs.

**Comment by svenklemm** [comment]
Wouldnt it be better to decouple the baserel cache from the hypertable cache

**Comment by mkindahl** [comment]
> Wouldnt it be better to decouple the baserel cache from the hypertable cache

Yes, but this is more work and we want to get the 2.9.0 release out. I was planning on refactoring this code to decouple the caches.

**Pull request merged**